### PR TITLE
Du: Implement du

### PR DESCRIPTION
**Pull request by KaindlJulian**
Implement du #23 

__References:__
- https://www.gnu.org/software/coreutils/manual/html_node/du-invocation.html
- https://www.gnu.org/software/coreutils/manual/html_node/Block-size.html#Block-size

__Missing flags__
- `--files0-from=file`
- `--exclude-from=file`

### CODE REVIEW

File: du/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+chrono = "^0.4.9"
+clap = { version = "^2.33.0", features = ["yaml", "wrap_help"] }
+walkdir = "^2.2.9"
+glob = "^0.3.0"
+
+[build-dependencies]
+chrono = "^0.4.9"
+clap = { version = "^2.33.0", features = ["yaml"] }
+walkdir = "^2.2.9"
+glob = "^0.3.0"
```
Comment (GrayJack) [inline du/Cargo.toml]: I think You just need clap for build.rs

File: du/src/du.yml [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+          multiple: true
+    - line-end-null:
+          long: 'null'
+          short: '0'
+          help: Ends each output line with NUL, not newline
+    - all:
+          long: all
+          short: a
+          help: Prints counts for all files, not just directories
+    - apparent-size:
```
Comment (GrayJack) [inline du/src/du.yml]: All itens should have long and short options, if in the original doesn't have one, fell free to create

File: du/src/du.yml [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+          value_name: SIZE
+    - bytes:
+          long: bytes
+          short: b
+          help: Equivalent to '--apparent-size --block-size=1'
+    - total:
+          long: total
+          short: c
+          help: Produces a grand total
+    - dereference-args:
```
Comment (GrayJack) [inline du/src/du.yml]: Missed the help msg

File: du/src/du.yml [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+          long: dereference-args
+          short: D
+    - max-depth:
+          long: max-depth
+          short: d
+          help: Prints the total for a directory only if it is N or fewer levels below the FILEs
```
Comment (GrayJack) [inline du/src/du.yml]: Always use "display" over "print"/"prints", for consistency with other utils.

**Comment by KaindlJulian** [comment]
Sry was a bit busy today. I fixed everything you mentioned. I also noticed that I forgot the `--total` flag. Will take care of that tomorrow :smile:

File: du/src/du.yml (hunk 2 above)
Comment (GrayJack) [reply du/src/du.yml]: Looks like on FreeBSD that one flag is `-A`

File: du/src/blocksize.rs [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+    PB,
+    P,
+    PiB,
+    EB,
+    E,
+    EiB,
+}
+
+impl BlocksizeSuffix {
+    fn iterate() -> Iter<'static, BlocksizeSuffix> {
```
Comment (GrayJack) [inline du/src/blocksize.rs]: Change name to `iter` to follow Rust naming conventions

File: du/src/blocksize.rs [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+        let suffix = BlocksizeSuffix::largest_from_value(total, self.use_si);
+
+        Blocksize { value: total / suffix.map_or(1, |s| s.value()), suffix, use_si: self.use_si }
+    }
+
+    pub fn value(&self) -> u64 { self.value * self.suffix.map_or(1, |s| s.value()) }
+
+    pub fn human_readable(&self) -> String { format!("{}", self) }
+
+    pub fn get_suffix_str(&self) -> &'static str {
```
Comment (GrayJack) [inline du/src/blocksize.rs]: Change name to `suffix_str` to follow Rust naming conventions

File: du/src/main.rs [hunk 7]
```diff
@@ -0,0 +1,578 @@
...
+}
+
+#[derive(Debug)]
+enum DisplayValue {
+    INodes(u64),
+    DiskUsage(Blocksize),
+}
+
+impl DisplayValue {
+    fn get_size(&self) -> u64 {
```
Comment (GrayJack) [inline du/src/main.rs]: Change name to `size` to follow Rust naming conventions

**Comment by GrayJack** [comment]
I had the time to review the code again today, again I didn't see anything to worry about, I just noted some stuff not following naming convention in Rust

**Comment by KaindlJulian** [comment]
Fixed the mentioned function names, gonna add tests in the next few days :smile:

**Comment by KaindlJulian** [comment]
I added tests for argument parsing. Don't really know how to test the stuff that depends on file metadata. Do you have any ideas @GrayJack ?

**Comment by GrayJack** [comment]
I have no idea yet, I've been looking for something to tests these cases, if you're agree, I'll merge this and if you add more tests, make a new pull request

**Comment by KaindlJulian** [comment]
Yea sounds good :+1:

**Pull request merged**